### PR TITLE
Fix 825 : Added Rust Implementation of Neural Networks from Scratch 

### PR DESCRIPTION
**Pull request by Prigoistic**
# PR Details

## 📄 Description of Changes

This PR adds a **minimal, self-contained feed-forward neural network implementation in Rust** under `nn_model.rs`.

### **Key Highlights**

* Two-layer network architecture: `Input → ReLU(Linear) → Softmax(Linear)`
* Forward pass producing intermediate activations (`z1`, `a1`, `z2`, `a2`)
* Backward pass computing analytical gradients for **cross-entropy + softmax**
* Parameter update step with a learning rate
* Single-sample prediction helper returning **top class + probabilities**

This serves as an **educational and practical example** of building simple neural networks with `ndarray` in Rust — useful for learners and as a base for extending towards **batch training**, **additional layers**, and **optimizers**.

---

##  What Problem Does It Solve?

* Provides a clean **NN reference in Rust** with well-separated steps:

  * Initialization
  * Forward
  * Backward
  * Update
  * Predict
* Demonstrates **ndarray usage** for matrix operations and broadcasting.
* Can be integrated with sample datasets or extended to **mini-batch training** and **evaluation pipelines**.

---

##  Related Issues / Discussions

* General initiative: *Hacktoberfest neural/network algorithms collection*
* Addresses the “**Basic NN in Rust**” feature request or related discussions.
* Please link any specific issues here if relevant.

---

## Implementation Overview

**File:** `nn_model.rs`
**Dependencies:** `ndarray`, `rand`

### **Public API**

```rust
NeuralNetwork::new(input_size, hidden_size, output_size) -> Self
forward(&self, x: &Array1<f32>) -> (Array1<f32>, Array1<f32>, Array1<f32>, Array1<f32>)
backward(&self, x, y_true, z1, a1, a2) -> (dw1, db1, dw2, db2)
update(&mut self, dw1, db1, dw2, db2, lr)
predict_single(&self, x) -> (usize, Array1<f32>)
```

### **Activations**

* Hidden: **ReLU**
* Output: **Softmax** (with max-shift for numerical stability)

### **Loss / Gradients**

* Cross-Entropy with Softmax; output gradient simplified as `a2 - y_true`

---

### **Forward Pass**

```text
z1 = x · w1 + b1  
a1 = relu(z1)  
z2 = a1 · w2 + b2  
a2 = softmax(z2)
```

### **Backward Pass**

```text
dz2 = a2 - y_true  
dw2 = a1^T · dz2  
db2 = dz2  
da1 = w2 · dz2  
dz1 = da1 ⊙ relu'(z1)  
dw1 = x^T · dz1  
db1 = dz1
```

### **Update Step**

* Applies SGD-style updates using scalar `lr`

### **Prediction**

* `predict_single` returns `(argmax class, probability vector)`

> **Note:** Current implementation supports **single-sample inputs (`Array1`)**.
> Extending to **batched inputs (`Array2`)** is a natural next step.

---

### **Sanity Checks Performed**

* Forward pass runs without panic on typical input sizes.
* Softmax uses **max-shift** for numerical stability; probabilities sum ≈ 1.
* `predict_single` returns valid class index and matching probability vector.

### **Recommended Follow-up Tests**

* Unit tests:

  * Output shapes from forward/backward
  * Softmax sum to 1
  * ReLU and derivative edge cases
  * Gradient check (numerical approximation)
*  Integration test:

  * Tiny dataset (e.g., XOR or 3-class toy) trained for a few iterations to confirm **loss decrease**.

---

##  Performance / Design Considerations

* Uses `ndarray` operations with broadcasting — ideal for small demos or learning.
* Single-sample code path improves clarity.
* For performance:

  * Mini-batch (`Array2`) inputs
  * Vectorized gradient aggregation
  * Advanced optimizers (Adam, Momentum)
  * Weight initialization (Xavier / He)

---

##  How to Use

```rust
// Construct the model
let mut nn = NeuralNetwork::new(input_size, hidden_size, output_size);

// Forward pass
let (z1, a1, z2, a2) = nn.forward(&x);

// Compute gradients
let (dw1, db1, dw2, db2) = nn.backward(&x, &y_true, &z1, &a1, &a2);

// Update parameters
nn.update(&dw1, &db1, &dw2, &db2, lr);

// Predict
let (class_idx, probs) = nn.predict_single(&x);
```

**Cargo.toml**

```toml
ndarray = "0.15"
rand = "0.8"
```

---

## Contributor Info

**Name:**
[Priyam Ghosh]

---

## Checklist Before Submitting

* [x] Code checked for errors and adherence to project style
* [x] Read and followed contribution guidelines
* [x] Participated in or referenced related discussions/issues
* [x] Added contributor name above

---

## Additional Comments

* Can extend with **batch training**, **loss helpers**, and **main.rs training demo** in follow-up PR.
* If a preferred numeric type (`f64`) or style for `ndarray` usage exists, I can adapt accordingly.

Thanks for maintaining this great collaborative repo! 🚀

### CODE REVIEW

**Comment by Prigoistic** [comment]
@DhanushNehru can you please review and merge this
